topdown/sets.go

```diff
@@ -58,22 +58,28 @@ func builtinSetIntersection(a ast.Value) (ast.Value, error) {
 
 // builtinSetUnion returns the union of the given input sets
 func builtinSetUnion(a ast.Value) (ast.Value, error) {
+	// The set union logic here is duplicated and manually inlined on
+	// purpose. By lifting this logic up a level, and not doing pairwise
+	// set unions, we avoid a number of heap allocations. This improves
+	// performance dramatically over the naive approach.
+	result := ast.NewSet()
 
 	inputSet, err := builtins.SetOperand(a, 1)
 	if err != nil {
 		return nil, err
 	}
 
-	result := ast.NewSet()
-
 	err = inputSet.Iter(func(x *ast.Term) error {
-		n, err := builtins.SetOperand(x.Value, 1)
+		item, err := builtins.SetOperand(x.Value, 1)
 		if err != nil {
 			return err
 		}
-		result = result.Union(n)
+		item.Foreach(func(y *ast.Term) {
+			result.Add(y)
+		})
```

Comment:
[golf] Isn't this the same as

```suggestion
		item.Foreach(result.Add)
```

One of those few lucky moments when the signatures _just fit_ 😅

Reply:
Oooooh! I think you might be right! :smiley: I'll test that shortly.

Reply:
Comitted that exact suggestion.